### PR TITLE
fix(ingestion/bigquery): user exceeded quota for concurrent project.lists requests

### PR DESCRIPTION
**Pull request by shubhamjagtap639**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 1]
```diff
@@ -150,12 +150,22 @@ def get_query_result(self, query: str) -> RowIterator:
...
-
-                return [
-                    BigqueryProject(id=p.project_id, name=p.friendly_name)
-                    for p in projects
-                ]
+                projects: List[BigqueryProject] = []
+                page_token: Optional[str] = None
+                while True:
+                    projects_iterator = self.bq_client.list_projects(
+                        max_results=5, page_token=page_token
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:157]: why is max_results=5?

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 2]
```diff
@@ -150,12 +150,22 @@ def get_query_result(self, query: str) -> RowIterator:
     def get_projects(self) -> List[BigqueryProject]:
         with self.report.list_projects:
             try:
-                projects = self.bq_client.list_projects()
-
-                return [
-                    BigqueryProject(id=p.project_id, name=p.friendly_name)
-                    for p in projects
-                ]
+                projects: List[BigqueryProject] = []
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:153]: add a comment explaining why this was required

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py (hunk 1 above)
Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py]: Sorry, this should be 50 as API returns max 50 projects per page. Refer [link](https://cloud.google.com/bigquery/docs/reference/rest/v2/projects/list#query-parameters)

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py (hunk 2 above)
Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:154]: Yes, Let me add the comment.
The thing is we are not able to regenerate the error but from [doc](https://cloud.google.com/bigquery/quotas#api_request_quotas) we got to know two project.lists request is allowed per second and assumed that bigquery client method list_projects is not adding any limit in calling request.
Hence we are trying to add limit in requests per second externally,

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 3]
```diff
@@ -165,6 +170,9 @@ def get_projects(self) -> List[BigqueryProject]:
                     page_token = projects_iterator.next_page_token
                     if page_token is None:
                         break
+                    # Sleep of 30 sec to make limit of two requests per second
+                    time.sleep(30)
```
[the remark this answers is not in the document]
Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py]: Added

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 4]
```diff
@@ -157,21 +157,19 @@ def get_projects(self) -> List[BigqueryProject]:
                 # Also project.list returns max 50 projects per page.
                 # Assuming list_projects method internally not adding any limit in requests call,
                 # externally we are adding limit in requests call per second.
+                rate_limiter = RateLimiter(max_calls=2, period=1)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:160]:
```suggestion
                rate_limiter = RateLimiter(max_calls=2, period=60)
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 5]
```diff
@@ -157,21 +157,19 @@ def get_projects(self) -> List[BigqueryProject]:
...
-                    )
-                    projects.extend(
-                        [
-                            BigqueryProject(id=p.project_id, name=p.friendly_name)
-                            for p in projects_iterator
-                        ]
-                    )
+                    with rate_limiter:
+                        projects_iterator = self.bq_client.list_projects(
+                            max_results=30, page_token=page_token
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:164]: what is the max number of results in a single page? can you add a link to the docs here?

Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:164]: Its 50 projects per page
https://cloud.google.com/bigquery/docs/reference/rest/v2/projects/list#query-parameters

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py (hunk 4 above)
Comment (shubhamjagtap639) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:160]: Quata for projects.list is 2 requests per second not per minute.
I know earlier I added the sleep of 30 sec but it was incorrect.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 6]
```diff
@@ -149,6 +150,10 @@ def get_query_result(self, query: str) -> RowIterator:
         return resp.result()
 
     def get_projects(self) -> List[BigqueryProject]:
+        def _should_retry(exc: BaseException) -> bool:
+            logger.debug(f"Exception reason: {str(exc)}. Type: {type(exc)}")
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:154]: add a comment here explaining what we previously tried and how we arrived at this solution

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 7]
```diff
@@ -148,14 +150,38 @@ def get_query_result(self, query: str) -> RowIterator:
...
-                projects = self.bq_client.list_projects()
-
-                return [
-                    BigqueryProject(id=p.project_id, name=p.friendly_name)
-                    for p in projects
-                ]
+                projects: List[BigqueryProject] = []
+                page_token: Optional[str] = None
+                # Bigquery API has limit in calling project.list request i.e. 2 request per second.
+                # Also project.list returns max 50 projects per page.
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:162]: add a link to the bigquery docs

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 8]
```diff
@@ -148,14 +150,38 @@ def get_query_result(self, query: str) -> RowIterator:
...
+                # Assuming list_projects method internally not adding any limit in requests call,
+                # externally we are adding limit in requests call per second.
+                rate_limiter = RateLimiter(max_calls=1, period=2)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:165]: this is 1 request per 2 seconds, not 2 requests per second?

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 9]
```diff
@@ -148,14 +150,38 @@ def get_query_result(self, query: str) -> RowIterator:
...
+                while True:
+                    with rate_limiter:
+                        projects_iterator = self.bq_client.list_projects(
+                            max_results=50,
+                            page_token=page_token,
+                            retry=retry.Retry(predicate=_should_retry, timeout=20),
+                        )
+                        count = 0
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:173]: make this more descriptive

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 10]
```diff
@@ -148,14 +150,38 @@ def get_query_result(self, query: str) -> RowIterator:
...
+                        for p in projects_iterator:
+                            projects.append(
+                                BigqueryProject(id=p.project_id, name=p.friendly_name)
+                            )
+                            count += 1
+                        logger.debug(f"Projects count per request {str(count)}")
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:179]: no need for str

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**